guacamole-ext/src/main/java/org/apache/guacamole/properties/TimeZoneGuacamoleProperty.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.properties;
+
+import java.util.TimeZone;
+import org.apache.guacamole.GuacamoleException;
+
+/**
+ * A GuacamoleProperty whose value is a TimeZone.
+ */
+public abstract class TimeZoneGuacamoleProperty
+        implements GuacamoleProperty<TimeZone> {
+    
+    @Override
+    public TimeZone parseValue(String value) throws GuacamoleException {
+        
+        // Nothing in, nothing out
+        if (value == null || value.isEmpty())
+            return null;
+        
+        // Attempt to return the TimeZone of the provided string value.
+        return TimeZone.getTimeZone(value);
```

Comment:
What if the provided timezone name is invalid?

Reply:
According to the `TimeZone` documentation, it simply returns GMT if the provided value is invalid.  Guess the Java developers were tired that day and decided not to use any Exception handling :laughing:.  Anyway, I can either document it as such, or throw in some logic that tries to detect when that happens (GMT is not provided as the string value, but ends up as the `TimeZone` value) and throw an exception.

Reply:
Yikes, that is a bit of a horror show: https://stackoverflow.com/questions/13092865/timezone-validation-in-java

If possible, I think proper validation is worth doing ... but I get that it's going to be real hairy. Would probably need a few tests that verify the validation works correctly for legal ways of specifying "GMT".

I hate to ask, but is there a finite, known set of strings for GMT within Java?

Reply:
> I hate to ask, but is there a finite, known set of strings for GMT within Java?

Judging from [the grammar in the documentation for `TimeZone`](https://docs.oracle.com/javase/7/docs/api/java/util/TimeZone.html), it looks like valid GMT strings are:

* `GMT`
* `GMT+0`
* `GMT-0`
* `GMT+0:00`
* `GMT-0:00`
* `GMT+000`
* `GMT-000`
* `GMT+00`
* `GMT-00`
* `GMT+00:00`
* `GMT-00:00`
* `GMT+0000`
* `GMT-0000`

We might be able to safely validate by checking that either of the following is true:

1. The resulting `TimeZone` is not equivalent to `GMT`.
2. The resulting `TimeZone` _is_ equivalent to `GMT`, and is either explicitly listed by `getAvailableIDs()` or matches a valid "custom time zone ID" for GMT (regex?).

Reply:
Is there any reason not to just loop through the `getAvailableIDs()` output and compare the string value we receive to the available IDs, and throw an error if don't match on that? It's probably a slightly expensive operation from a compute perspective, but probably not all that bad, and then we can just document that the value for the timezone property must be a valid Java TimeZone ID. Something like:
```
for (String tzStr : TimeZone.getAvailableIDs()) {
    if (value.equals(tzStr))
        return TimeZone.getTimeZone(value);
}

throw GuacamoleServerException("Invalid timezone specified.");
```

It's not pretty or elegant, but it's fairly simple. Or is there some downfall I'm missing with that?

Reply:
I believe so: not _all_ valid GMT representations are within that list. The grammar for "custom time zone IDs" allows for quite a few equivalent ways to say GMT / GMT+0.

I think the best way would to accept any value that does not equate to GMT, and for all others do a quick comparison for known-good values (static `HashSet` of `getAvailableIDs()`?) and known-good representations of GMT (regex?). Anything that equals GMT, isn't a valid custom ID for GMT, and isn't explicitly listed by `getAvailableIDs()` must be invalid.

Reply:
Okay, I think I understand - will give it a go.

Reply:
Wow, the Java `TimeZone` implementation is really awful.  Writing the `JUnit` tests is fun:
* If you use the string "GMT" for the input, the ID you get back is "GMT" - okay, good so far.
* If you use the string "GMT+00:00" for the input, the ID you get back is "GMT+00:00" - not ideal, but not the worst.
* If you use the string "GMT+0000" for the input, the ID you get back is "GMT+00:00" - :exploding_head: 

So, cannot check that valid input strings generate an ID of "GMT", and cannot check that valid input strings generate an ID equal to the input string.

Reply:
Okay, well, I have given it a shot, so let me know how that looks.

Reply:
I feel like there might be two RegEx matches required, here - one for the valid GMT inputs, and one for TimeZone IDs that are returned for either GMT or custom offsets.  It seems that any time an input value is in the format "GMT+hhmm" it gets converted with `getID()` to "GMT+hh:mm" (adds the colon).  So, it seems like the checks for validity should be:
* Matches the list of known TimeZones
* Input value is the same as the `getID()` output
* Valid GMT value (GMT+/- and some number of 0s)
* Some sort of RegEx that matches custom offset of GMT[+-][hhmm] to GMT[+-][hh:mm]

Reply:
What about something like `^GMT[+-](0|00)((:)?00)?$` ?

Reply:
Yeah, one slight modification - since "GMT" is also valid, had to make the entire section after that optional - but that works pretty well.

Reply:
Strictly speaking, it isn't valid for a "custom time zone ID". From [the documentation for `TimeZone`](https://docs.oracle.com/javase/7/docs/api/java/util/TimeZone.html):

> You can use the getAvailableIDs method to iterate through all the supported time zone IDs. You can then choose a supported ID to get a TimeZone. If the time zone you want is not represented by one of the supported IDs, then a custom time zone ID can be specified to produce a TimeZone. The syntax of a custom time zone ID is:
>
> ```
>  CustomID:
>          GMT Sign Hours : Minutes
>          GMT Sign Hours Minutes
>          GMT Sign Hours
>  Sign: one of
>          + -
>  Hours:
>          Digit
>          Digit Digit
>  Minutes:
>          Digit Digit
>  Digit: one of
>          0 1 2 3 4 5 6 7 8 9
> ```

"GMT" is a valid time zone ID, but because `TimeZone` explicitly accepts it as one of the various supported IDs defined by `getAvailableIDs()`.

Reply:
So, are you good with the code as-is, or would you rather see a loop through `getAvailableIDs()` and put the RegEx back to requiring `GMT[+-]0`?